auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,196 @@
+        win(auction_id);
+    }
+
+    fn setup_auction(opening_price: u64, reserve_price: u64, start_time: u64, end_time: u64) -> u64 {
+        require(storage.constructed == true, Error::ContractNotConstructedYet);
+
+        require(get_sender() == storage.admin, Error::SenderNotAdmin);
+        require(opening_price > reserve_price, Error::EndPriceCannotBeLargerThanStartPrice);
+        require(end_time > height(), Error::AuctionCannotEndInThePast);
+        require(start_time >= height(), Error::AuctionCannotStartInThePast);
+        require(end_time > start_time, Error::AuctionCannotEndBeforeItStarts);
+
+        storage.latest_auction_id = storage.latest_auction_id + 1;
+        let current_auction_id = storage.latest_auction_id;
+
+        let mut auction = storage.auctions.get(current_auction_id);
+        auction.opening_price = opening_price;
+        auction.reserve_price = reserve_price;
+        auction.start_time = start_time;
+        auction.end_time = end_time;
+        auction.ended = false;
+        
+        storage.auctions.insert(current_auction_id, auction);
+
+        /// Returns the auction being setup's auction_id
+        current_auction_id
+    }
+
+    fn end_auction(auction_id: u64) {
+        require(storage.constructed == true, Error::ContractNotConstructedYet);
+
+        /// Only the admin can end the auction (prematurely)
+        require(get_sender() == storage.admin, Error::SenderNotAdmin);
+
+        /// If there is no auction going on currently the ended value will already be true so no need to check for that case
+        let mut auction = storage.auctions.get(auction_id);
+        auction.ended = true;
+        storage.auctions.insert(auction_id, auction);
+    }
+
+}
+
+fn win(auction_id: u64) {
+    ///  Do stuff on a winning bid, this function is called whenever a winning bid is recieved.
+
+    /// Add whatever logic you may want to execute on a win
+
+    //Currently just sends the bid amount to the beneficiary
+    let auction = storage.auctions.get(auction_id);
+    transfer_to_output(calculate_price(auction_id), storage.asset_id, auction.beneficiary);
+}
+
+fn calculate_price(auction_id: u64) -> u64 {
+    let auction = storage.auctions.get(auction_id);
+    
+    /// How much the price will go down by, throughout the auction
+    let price_delta = auction.opening_price - auction.reserve_price;
+    /// How long the auction will last
+    let auction_duration = auction.end_time - auction.start_time;
+    /// This is the amount the price will reduce by per block
+    let price_shift = price_delta / auction_duration;
+
+    /// Tells us how far we are into the auction (out of the auction_duration)
+    let now = height() - auction.start_time; 
+
+    /// Cap how far we are into the auction by the auction_duration, so price doesnt go into negative or below endprice
+    let now = if now > auction_duration { auction_duration } else { now };
```

Comment:
```suggestion
    let time_passed = if time_passed > auction_duration { auction_duration } else { time_passed };
```

Reply:
I'll use "blocks_into_auction" if thats fine instead, seems less confusing than either of those